ras-server.c

```diff
@@ -0,0 +1,363 @@
+		log(ALL, LOG_WARNING, "Can't create server thread\n");
+		goto create_err;
+	}
+
+	signal(SIGPIPE, SIG_IGN);
+	log(ALL, LOG_INFO, "Server is listening to connections\n");
+	return 0;
+
+	create_err:
+	ras_server_stop();
+	return -1;
+}
+
+// Serialization function for MC events
+static inline void ras_mc_msg(char *buf, struct ras_mc_event *ev) {
+	snprintf(buf, SRV_MSG_SIZE, "type=mc,"
+		"timestamp=%s,"
+		"error_count=%d,"
+		"error_type=%s,"
+		"msg=%s,"
+		"label=%s,"
+		"mc_index=%c,"
+		"top_layer=%c,"
+		"middle_layer=%c,"
+		"lower_layer=%c,"
+		"address=%llu,"
+		"grain=%llu,"
+		"syndrome=%llu,"
+		"driver_detail=%s",
+		ev->timestamp,
+		ev->error_count,
+		ev->error_type,
+		ev->msg,
+		ev->label,
+		ev->mc_index,
+		ev->top_layer,
+		ev->middle_layer,
+		ev->lower_layer,
+		ev->address,
+		ev->grain,
+		ev->syndrome,
+		ev->driver_detail);
+}
+
+// Serialization function for AER events
+static inline void ras_aer_msg(char *buf, struct ras_aer_event *ev) {
+	snprintf(buf, SRV_MSG_SIZE, "type=aer,"
+		"timestamp=%s,"
+		"error_type=%s,"
+		"dev_name=%s,"
+		"msg=%s",
+		ev->timestamp,
+		ev->error_type,
+		ev->dev_name,
+		ev->msg);
+}
+
+// Serialization function for MCE events
+static inline void ras_mce_msg(char *buf, struct mce_event *ev) {
+	snprintf(buf, SRV_MSG_SIZE, "type=mce,"
+		"timestamp=%s,"
+		"bank_name=%s,"
+		"mc_location=%s,"
+		"error_msg=%s",
+		ev->timestamp,
+		ev->bank_name,
+		ev->mc_location,
+		ev->error_msg);
+}
+
+// Serialization function for non_standard events
+static inline void ras_non_standard_msg(char *buf,
+										struct ras_non_standard_event *ev) {
+	snprintf(buf, SRV_MSG_SIZE, "type=non_standard,"
+		"timestamp=%s,"
+		"severity=%s,"
+		"length=%d",
+		ev->timestamp,
+		ev->severity,
+		ev->length);
+}
+
+// Serialization function for ARM events
+static inline void ras_arm_msg(char *buf, struct ras_arm_event *ev) {
+	snprintf(buf, SRV_MSG_SIZE, "type=arm,"
+		"timestamp=%s,"
+		"error_count=%d,"
+		"affinity=%d,"
+		"mpidr=0x%lx,"
+		"midr=0x%lx,"
+		"running_state=%d,"
+		"psci_state=%d",
+		ev->timestamp,
+		ev->error_count,
+		ev->affinity,
+		ev->mpidr,
+		ev->midr,
+		ev->running_state,
+		ev->psci_state);
+}
+
+// Serialization function for devlink events
+static inline void ras_devlink_msg(char *buf, struct devlink_event *ev) {
+	snprintf(buf, SRV_MSG_SIZE, "type=devlink,"
+		"timestamp=%s,"
+		"bus_name=%s,"	
+		"dev_name=%s,"	
+		"driver_name=%s,"
+		"reporter_name=%s,"
+		"msg=%s",
+		ev->timestamp,
+		ev->bus_name,
+		ev->dev_name,
+		ev->driver_name,
+		ev->reporter_name,
+		ev->msg);
+}
+
+// Serialization function for diskerror events
+static inline void ras_diskerror_msg(char *buf, struct diskerror_event *ev) {
+	snprintf(buf, SRV_MSG_SIZE, "type=diskerror,"
+		"timestamp=%s,"
+		"dev=%s,"
+		"sector=%llu,"
+		"nr_sector=%u,"
+		"error=%s,"
+		"rwbs=%s,"
+		"cmd=%s",
+		ev->timestamp,
+		ev->dev,
+		ev->sector,
+		ev->nr_sector,
+		ev->error,
+		ev->rwbs,
+		ev->cmd);
+}
+
```

Comment:
I'm a little bit worried about maintainership with the above. As newer messages and types are added at rasdaemon, the above will become outdated with time. IMO, the better here would be to have those feature-specific functions inside the code where they are added, in a way that, if something changes there, such change will be reflected at the broadcasting code. Granted, ras-report.c has the same issue, but it is at least a single place where we have those things. Now that a new report daemon is added, we should unify, in order to make easier to maintain it.

Reply:
Would a couple of functions to create strings from events be enough?
Something in the lines of:

```C
// Serialize a RAS event field by field, separated by 'sep'
void ras_to_str(int type, void* ev, char sep, char* buf);
// Get the name of a RAS event type
const char* ras_event_name(int type);
```
Is an error message on the default clause of a switch statement enough to warn developers of the need to serialize new event types?